Source/data/iterators.hpp

```diff
@@ -1,234 +1,408 @@
+		if (state_->endOfRecord()) {
+			state_ = nullptr;
+		} else {
+			unsigned fieldsSkipped = 0;
+			// By this point we've already advanced past the end of this field (either because the
+			//  last value access found the end of the field by necessity or we discarded it a few
+			//  lines up), so we only need to advance further if an increment greater than 1 was
+			//  provided.
+			*state_ = DiscardMultipleFields(state_->next, end_, increment - 1, &fieldsSkipped);
+			// As we've consumed the current field by this point we need to increment the internal
+			//  column counter one extra time so we have an accurate value.
+			column_ += fieldsSkipped + 1;
+			// We use Status::ReadyToRead as a marker so we only read the next value on the next
+			//  value access, this allows consumers to choose the most efficient method (e.g. if
+			//  they want the value as an int) or even repeated advances without using a value.
+			state_->status = GetFieldResult::Status::ReadyToRead;
 		}
+		return *this;
+	}
 
-		/**
-		 * @brief Checks if this field is the last field in a file, not just in the record
-		 *
-		 * If you call this method before calling parseInt or operator* then this will trigger a
-		 * read, meaning you can no longer use parseInt to try extract a numeric value. You
-		 * probably want to use this after calling parseInt.
-		 * @return true if this field is the last field in the file/RecordsRange
-		 */
-		[[nodiscard]] bool endOfFile() const
-		{
-			if (state_->status == GetFieldResult::Status::ReadyToRead) {
-				*state_ = GetNextField(state_->next, end_);
-			}
-			return state_->endOfFile();
-		}
-	};
+	/**
+	 * @brief Returns a view of the current field
+	 *
+	 * The returned value is a thin wrapper over the current state of this iterator (or last
+	 * successful read if incrementing this iterator would result in it reaching the end state).
+	 */
+	[[nodiscard]] value_type operator*()
+	{
+		return { state_, end_, row_, column_ };
+	}
 
-	[[nodiscard]] Iterator begin() const
+	/**
+	 * @brief Checks if this field is the last field in a file, not just in the record
+	 *
+	 * If you call this method before calling Field::parseInt or Field::operator* then this will
+	 * trigger a read, it's more efficient to call this after calling parseInt.
+	 * @return true if this field is the last field in the file
+	 */
+	[[nodiscard]] bool endOfFile() const
 	{
-		return { state_, end_ };
+		if (state_->status == GetFieldResult::Status::ReadyToRead) {
+			*state_ = GetNextField(state_->next, end_);
+		}
+		return state_->endOfFile();
 	}
 
-	[[nodiscard]] Iterator end() const
+	/**
+	 * @brief Returns the current row number
+	 */
+	[[nodiscard]] unsigned row() const
 	{
-		return {};
+		return row_;
+	}
+	/**
+	 * @brief Returns the current column/field number (from the start of the row/record)
+	 */
+	[[nodiscard]] unsigned column() const
+	{
+		return column_;
 	}
 };
 
-class RecordsRange {
-	const char *const begin_;
+class Record {

```

Comment:
The names `Record` and `Field` in the `devilution` namespace are too general.
How about `DataFileRecord`, `DataFileField`?

Reply:
Definitely agree there. Would it be reasonable to declare them as inner classes of `DataFile` so they get accessed as `DataFile::Record` etc? Or introduce another namespace like `devilution::textdata::Record`?

Reply:
Yeah either of these options are fine